src/libraries/System.Linq/src/System/Linq/Single.cs

```diff
@@ -10,82 +10,46 @@ public static partial class Enumerable
     {
         public static TSource Single<TSource>(this IEnumerable<TSource> source)
         {
-            if (source == null)
+            var single = source.TryGetSingle(out bool found);
+            if (!found)
             {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
-            }
-
-            if (source is IList<TSource> list)
-            {
-                switch (list.Count)
-                {
-                    case 0:
-                        ThrowHelper.ThrowNoElementsException();
-                        return default;
-                    case 1:
-                        return list[0];
-                }
-            }
-            else
-            {
-                using (IEnumerator<TSource> e = source.GetEnumerator())
-                {
-                    if (!e.MoveNext())
-                    {
-                        ThrowHelper.ThrowNoElementsException();
-                    }
-
-                    TSource result = e.Current;
-                    if (!e.MoveNext())
-                    {
-                        return result;
-                    }
-                }
+                ThrowHelper.ThrowNoElementsException();
             }
 
-            ThrowHelper.ThrowMoreThanOneElementException();
-            return default;
+            return single!;
         }
-
         public static TSource Single<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
-            if (source == null)
+            var single = source.TryGetSingle(predicate, out bool found);
+            if (!found)
             {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+                ThrowHelper.ThrowNoElementsException();
```

Comment:
The predicate overload throws a different exception:
```suggestion
                ThrowHelper.ThrowNoMatchException();
```